File: Data Structures & Algorithms/permutation-string/submission-1.cpp
class Solution {
public:
    bool checkInclusion(string s1, string s2) {
        if (s1.length() > s2.length()) return false;

        vector<int> freq1(26, 0), freq2(26, 0);

        // Count frequency of s1
        for (char c : s1) {
            freq1[c - 'a']++;
        }

        // Sliding window on s2
        for (int i = 0; i < s2.length(); i++) {
            freq2[s2[i] - 'a']++;  // Add new character to window

            // Remove character left out of the window
            if (i >= s1.length()) {
                freq2[s2[i - s1.length()] - 'a']--;
            }

            // Compare frequency arrays
            if (freq1 == freq2) return true;
        }

        return false;
    }
};

File: Data Structures & Algorithms/minimum-window-with-characters/submission-1.java
class Solution {
    public String minWindow(String s, String t) {

        if (s.length() < t.length()) {
            return "";
        }

        int[] count = new int[128];

        // Store frequency of t
        for (char c : t.toCharArray()) {
            count[c]++;
        }

        int left = 0;
        int required = t.length();

        int minLen = Integer.MAX_VALUE;
        int start = 0;

        for (int right = 0; right < s.length(); right++) {

            char current = s.charAt(right);

            // Needed character found
            if (count[current] > 0) {
                required--;
            }

            count[current]--;

            // Valid window found
            while (required == 0) {

                if (right - left + 1 < minLen) {
                    minLen = right - left + 1;
                    start = left;
                }

                char leftChar = s.charAt(left);

                count[leftChar]++;

                if (count[leftChar] > 0) {
                    required++;
                }

                left++;
            }
        }

        return minLen == Integer.MAX_VALUE
                ? ""
                : s.substring(start, start + minLen);
    }
}